src/backend/curses/pan.rs

```diff
@@ -0,0 +1,219 @@
+extern crate pancurses;
+
+use backend;
+use event::{Event, Key};
+
+use self::super::find_closest;
+use theme::{BaseColor, Color, ColorStyle, Effect};
+use utf8;
+
+pub struct Concrete {
+    window: pancurses::Window,
+}
+
+impl backend::Backend for Concrete {
+    fn init() -> Self {
+        ::std::env::set_var("ESCDELAY", "25");
+        let window = pancurses::initscr();
+        window.keypad(true);
+        pancurses::noecho();
+        pancurses::cbreak();
+        pancurses::start_color();
+        pancurses::curs_set(0);
+        window.bkgd(pancurses::COLOR_PAIR(ColorStyle::Background.id() as u64));
+
+        Concrete { window: window }
+    }
+
+    fn screen_size(&self) -> (usize, usize) {
+        let (x, y) = self.window.get_max_yx();
+        (x as usize, y as usize)
+    }
+
+    fn has_colors(&self) -> bool {
+        pancurses::has_colors()
+    }
+
+    fn finish(&mut self) {
+        pancurses::endwin();
+    }
+
+    fn init_color_style(&mut self, style: ColorStyle, foreground: &Color,
+                        background: &Color) {
+        pancurses::init_pair(style.id(),
+                             find_closest(foreground) as i16,
+                             find_closest(background) as i16);
+    }
+
+    fn with_color<F: FnOnce()>(&self, color: ColorStyle, f: F) {
+        // TODO: pancurses doesn't have an `attr_get` equivalent
+        // let mut current_style: pancurses::attr_t = 0;
+        // let mut current_color: i16 = 0;
+        // pancurses::attr_get(&mut current_style, &mut current_color);
+
+        let style = pancurses::COLOR_PAIR(color.id() as u64);
+        self.window.attron(style);
+        f();
+        self.window.attroff(style);
+        // pancurses::attron(current_style);
+    }
+
+    fn with_effect<F: FnOnce()>(&self, effect: Effect, f: F) {
+        let style = match effect {
+            // A_REVERSE, taken from ncurses
+            Effect::Reverse => 1 << (10 + 8u32),
+            Effect::Simple => pancurses::A_NORMAL,
+        };
+        self.window.attron(style);
+        f();
+        self.window.attroff(style);
+    }
+
+    fn clear(&self) {
+        self.window.clear();
+    }
+
+    fn refresh(&mut self) {
+        self.window.refresh();
+    }
+
+    fn print_at(&self, (x, y): (usize, usize), text: &str) {
+        self.window.mvaddstr(y as i32, x as i32, text);
+    }
+
+    fn poll_event(&self) -> Event {
+        // TODO: there seems to not be any indication
+        // of Ctrl/Alt/Shift in these :v
```

Comment:
For ncurses, a lot of the key+modifiers code were undocumented, outside of the usual code range, and I had to manually try every key combination, write down the bytes received by the application, and use this as mapping.